lib/formatters/sarif.js

```diff
@@ -0,0 +1,182 @@
+                  id: message.rule,
+                  helpUri: `https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/${message.rule}.md`,
+                };
+              }
+
+              if (sarifRuleIndices[message.rule] !== 'undefined') {
+                sarifRepresentation.ruleIndex = sarifRuleIndices[message.rule];
+              }
+            } else {
+              // ember-template-lint produces a message with rule set to 'global' when it
+              // encounters an internal error. SARIF represents this as a tool execution
+              // notification rather than as a result, and a notification has a descriptor.id
+              // property rather than a rule property.
+              sarifRepresentation.descriptor = {
+                id: 'ETL-NO-RULEID',
+              };
+            }
+
+            if (message.line > 0 || message.column > 0) {
+              sarifRepresentation.locations[0].physicalLocation.region = {};
+              if (message.line > 0) {
+                sarifRepresentation.locations[0].physicalLocation.region.startLine = message.line;
+              }
+              if (message.column > 0) {
+                sarifRepresentation.locations[0].physicalLocation.region.startColumn =
+                  message.column;
+              }
+            }
+
+            if (message.source) {
+              // Create an empty region if we don't already have one from the line / column block above.
+              sarifRepresentation.locations[0].physicalLocation.region =
+                sarifRepresentation.locations[0].physicalLocation.region || {};
+              sarifRepresentation.locations[0].physicalLocation.region.snippet = {
+                text: message.source,
+              };
+            }
+
+            if (message.rule) {
+              sarifResults.push(sarifRepresentation);
+            } else {
+              toolConfigurationNotifications.push(sarifRepresentation);
+            }
+          }
+        }
+      }
+    }
+
+    if (Object.keys(sarifFiles).length > 0) {
+      sarifLog.runs[0].artifacts = [];
+
+      for (const path of Object.keys(sarifFiles)) {
+        sarifLog.runs[0].artifacts.push(sarifFiles[path]);
+      }
+    }
+
+    // Per the SARIF spec §3.14.23, run.results must be present even if there are no results.
+    // This provides a positive indication that the run completed and no results were found.
+    sarifLog.runs[0].results = sarifResults;
+
+    if (toolConfigurationNotifications.length > 0) {
+      sarifLog.runs[0].invocations = [
+        {
+          toolConfigurationNotifications,
+          executionSuccessful: true,
+        },
+      ];
+    }
+
+    if (Object.keys(sarifRules).length > 0) {
+      for (const ruleId of Object.keys(sarifRules)) {
+        let rule = sarifRules[ruleId];
+        sarifLog.runs[0].tool.driver.rules.push(rule);
+      }
+    }
+
+    this.console.log(JSON.stringify(sarifLog, null, 2));
```

Comment:
I'd prefer to have this save to a file, instead of `console.log` (otherwise you interleave any arbitrary stdout along with the actual sarif output and end up with something that is unparsable).

Reply:
We can do that. I can do something similar to what we do with Checkup.

Reply:
If we do this, it seems like we should provide an `--output-file` option too, unless we just want to autogenerate SARIF log names. Thoughts?

Reply:
Ya, agree RE: adding `--output-file`. In addition we could choose a reasonable filename by default (to make `ember-template-lint . --format sarif` work without an `--output-path`).

Reply:
Agreed.

Reply:
Oh, another factor here is that we should emit to stdout when piping.

So I think the decision tree here is:

* if `--output-file` is provided, use it and save a file
* if piping, log the content to stdout
* choose a filename and save to working dir

Reply:
OK, I've added the functionality to pipe to stdout.